[GHF] Print better merge rule match reason

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74887
* #74893
* #74892

When running `find_matching_merge_rule` raise an exception that:
 - Contains generic message if files changed do not match any rules
 - Mentions rule that matches majority of files, but calls out few
   non-matching ones
 - Mentions matching rules, but calls out that it lacks reviewers
 - Mentions rule, but calls out missing mandatory checks

malfet committed:
[GHF] Print better merge rule match reason

When running `find_matching_merge_rule` raise an exception that:
 - Contains generic message if files changed do not match any rules
 - Mentions rule that matches majority of files, but calls out few
   non-matching ones
 - Mentions matching rules, but calls out that it lacks reviewers
 - Mentions rule, but calls out missing mandatory checks

[ghstack-poisoned]

janeyx99 commented on .github/scripts/trymerge.py:
```diff
@@ -579,46 +578,58 @@ def read_merge_rules(repo: GitRepo) -> List[MergeRule]:
+    # Score 0 to 10K - how many files rule matched
+    # Score 10K - matched all files, but no overlapping approvers
+    # Score 20K - matched all files and approvers, but lacks mandatory checks
+    reject_reason_score = 0
```
```suggestion
    num_matching_files = 0
```
reject_reason_score is a confusing name here

janeyx99 replied:
oh nvm i see

janeyx99 commented on .github/scripts/trymerge.py:
```diff
@@ -579,46 +578,58 @@ def read_merge_rules(repo: GitRepo) -> List[MergeRule]:
+            if not patterns_re.match(fname):
+                non_matching_files.append(fname)
+        if len(non_matching_files) > 0:
+            score = len(changed_files) - len(non_matching_files)
```
```suggestion
            num_matching_files = len(changed_files) - len(non_matching_files)
```
score is a vague variable--this is a nit

malfet committed:
Update on "[GHF] Print better merge rule match reason"


When running `find_matching_merge_rule` raise an exception that:
 - Contains generic message if files changed do not match any rules
 - Mentions rule that matches majority of files, but calls out few
   non-matching ones
 - Mentions matching rules, but calls out that it lacks reviewers
 - Mentions rule, but calls out missing mandatory checks

[ghstack-poisoned]

malfet commented:
Applied both suggestions

osalpekar reviewed (approved):
Looks great - thanks!

malfet commented:
@pytorchbot merge this
